### PR TITLE
fix memory leak in ShapeManager

### PR DESCRIPTION
**Pull request by AndrewMeadows**
- fix memory leak in ShapeManager

### CODE REVIEW

**Comment by AndrewMeadows** [comment]
https://app.asana.com/0/26623005675174/103442220436762

@sethalves noticed that there was a memory leak in the ShapeManager.  The problem was that the child shapes of "compound shapes" are not yet in the ShapeManager, so we need to delete the children whenever we delete their parent.  Models that use collision hulls for their shape, and PolyVox entities use the "compound shape" feature.

**Comment by AndrewMeadows** [comment]
Assigning to @sethalves to test and review: he knows the testplan which goes something like this:

(1) build on linux with this environment variable defined: export HIFI_MEMORY_DEBUGGING=1
(2) delete all controller plugins
(3) run the interface and capture all logs (including stderr)
(4) navigate to a domain with lots of compound objects
(5) logout and examine the tail of the logs for memory leak information

**Comment by hifi-gustavo** [comment]
Build console output for commit: ecfe198e35fad5b07d09eba40ede017b7f1eb65c available at: https://gist.github.com/1b5e7d43cb5ecfbe7df5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7445.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7445.dmg

File: libraries/physics/src/ShapeManager.cpp [hunk 1]
```diff
@@ -58,14 +59,14 @@ btCollisionShape* ShapeManager::getShape(const ShapeInfo& info) {
...
-bool ShapeManager::releaseShape(const DoubleHashKey& key) {
+bool ShapeManager::releaseShapeByKey(const DoubleHashKey& key) {
     ShapeReference* shapeRef = _shapeMap.find(key);
     if (shapeRef) {
         if (shapeRef->refCount > 0) {
             shapeRef->refCount--;
             if (shapeRef->refCount == 0) {
                 _pendingGarbage.push_back(key);
-                const int MAX_GARBAGE_CAPACITY = 127;
+                const int MAX_GARBAGE_CAPACITY = 255;
```
Comment (AndrewMeadows) [inline libraries/physics/src/ShapeManager.cpp]: I'm bumping this up in a half-hearted attempt to reduce immediate garbage collection when deleting entities using compound shapes.  However, I noticed that many of our compound shapes are using 1000+ compound parts.  None of the child shapes are shared atm, unless their parent shape itself is shared.

File: libraries/physics/src/ShapeFactory.cpp [hunk 2]
```diff
@@ -120,3 +120,21 @@ btCollisionShape* ShapeFactory::createShapeFromInfo(const ShapeInfo& info) {
...
+            if (childShape->getShapeType() == (int)COMPOUND_SHAPE_PROXYTYPE) {
+                // recurse
+                ShapeFactory::deleteShape(childShape);
+            } else {
+                delete childShape;
+            }
+        }
+    }
+    delete shape;
+}
```
Comment (AndrewMeadows) [inline libraries/physics/src/ShapeFactory.cpp]: This is the real fix: recursive delete of child shapes.  Yes, I noticed at least one compound shape heirarchy that was two deep:  CompoundShape with one child: a CompoundShape.  This particular inefficiency should eventually be fixed but eventually we'll probably have CompoundShapes with many CompoundShape children (for articulated objects with complex parts).

**Comment by sethalves** [comment]
I've tested this, and it looks good.

**Pull request merged**